src/common/timing/time-keeper.js

```diff
@@ -0,0 +1,90 @@
+import { gosNREUM } from '../window/nreum'
+import { globalScope } from '../constants/runtime'
+
+/**
+ * Class used to adjust the timestamp of harvested data to New Relic server time. This
+ * is done by tracking the performance timings of the RUM call and applying a calculation
+ * to the harvested data event offset time.
+ */
+export class TimeKeeper {
+  /**
+   * Represents the browser origin time corrected to NR server time.
+   * @type {int}
+   */
+  #correctedOriginTime
+
+  /**
+   * Represents the page origin time offset of the NR server time. This is the difference
+   * between the local browser time and the NR server time.
+   * @type {int}
+   */
+  #serverOffset
+
+  static getTimeKeeperByAgentIdentifier (agentIdentifier) {
+    const nr = gosNREUM()
+    return Object.keys(nr?.initializedAgents || {}).indexOf(agentIdentifier) > -1
+      ? nr.initializedAgents[agentIdentifier].timeKeeper
+      : undefined
+  }
+
+  get correctPageOriginTime () {
+    return this.#correctedOriginTime
+  }
+
+  /**
+   * Process a rum request to calculate NR server time.
+   * @param rumRequest {XMLHttpRequest} The xhr for the rum request
+   * @param rumRequestUrl {string} The full url of the rum request
+   */
+  processRumRequest (rumRequest, rumRequestUrl) {
+    const responseDateHeader = rumRequest.getResponseHeader('Date')
+    if (!responseDateHeader) {
+      throw new Error('Missing date header on rum response.')
+    }
+
+    const resourceEntries = globalScope.performance.getEntriesByName(rumRequestUrl, 'resource')
+    if (!Array.isArray((resourceEntries)) || resourceEntries.length === 0) {
+      throw new Error('Missing rum request performance entry.')
+    }
+
+    let medianRumOffset = 0
+    this.#serverOffset = 0
+    if (typeof resourceEntries[0].responseStart === 'number' && resourceEntries[0].responseStart !== 0) {
+      // Cors is enabled and we can make a more accurate calculation of NR server time
+      medianRumOffset = (resourceEntries[0].responseStart - resourceEntries[0].requestStart) / 2
+      this.#serverOffset = Math.floor(resourceEntries[0].requestStart + medianRumOffset)
+    } else {
+      // Cors is disabled or erred, we need to use a less accurate calculation
+      medianRumOffset = (resourceEntries[0].responseEnd - resourceEntries[0].fetchStart) / 2
+      this.#serverOffset = Math.floor(resourceEntries[0].fetchStart + medianRumOffset)
+    }
+
+    // This is the NR server time
+    const serverTimestamp = Date.parse(responseDateHeader) - medianRumOffset
+    // Corrected page origin time
+    this.#correctedOriginTime = Math.floor(serverTimestamp - this.#serverOffset)
```

Comment:
I'm not understanding `serverOffset` or this calculation. So that I'm clear, an event that occur at some later time will have a relative DOMHighResTimeStamp that's based on client `timeOrigin`, different from `serverTimestamp` aka "server timeOrigin".
We want to map that old DOMHighResTimeStamp to a new DOMHighResTimeStamp that's relative to `serverTimestamp` at this particular point in time, right?
If so, why not capture the DOMHighResTimeStamp right here at this point in time, and _subtract_ it from the event's later DOMHighResTimeStamp -- let's call this `newPerfNow`? Then the "server time" at which the client side event occurred would simply be `serverTimestamp` + `newPerfNow`.

Reply:
[Here](https://whimsical.com/calculated-nr-time-5ihwHhG6VhcTGV9Czi6BfU) is the whimsical that details out the calculations a little more. Let me know if that and the above answers are still unclear.

Reply:
<img width="340" alt="image" src="https://github.com/newrelic/newrelic-browser-agent/assets/28714891/2f796074-79e6-4260-8ec0-fa41fb750ece">
`serverTimestamp` doesn't seem to exist in the whimsical; it uses the dateHeader parsed directly without subtracting the rumOffset. This in part threw me off because in the final equation, `medianRumOffset` is subtracted twice.